src/c++/perf_analyzer/metrics_manager.cc

```diff
@@ -0,0 +1,164 @@
+// are met:
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "metrics_manager.h"
+#include <stdexcept>
+
+namespace triton { namespace perfanalyzer {
+
+MetricsManager::MetricsManager(
+    std::shared_ptr<clientbackend::ClientBackend> client_backend,
+    uint64_t metrics_interval_ms)
+    : client_backend_(client_backend), metrics_interval_ms_(metrics_interval_ms)
+{
+}
+
+MetricsManager::~MetricsManager()
+{
+  if (query_loop_future_.valid()) {
+    StopQueryingMetrics();
+  }
+}
+
+void
+MetricsManager::StartQueryingMetrics()
+{
+  should_keep_querying_ = true;
+  query_loop_future_ =
+      std::async(&MetricsManager::QueryMetricsEveryNMilliseconds, this);
+}
+
+void
+MetricsManager::QueryMetricsEveryNMilliseconds()
+{
+  while (should_keep_querying_) {
+    const auto& start{std::chrono::system_clock::now()};
+
+    Metrics metrics{};
+    clientbackend::Error err{client_backend_->Metrics(metrics)};
+    if (err.IsOk() == false) {
+      throw std::runtime_error(err.Message());
+    }
+
+    CheckForMissingMetrics(metrics);
+
+    {
+      std::lock_guard<std::mutex> metrics_per_timestamp_lock{
+          metrics_per_timestamp_mutex_};
+      metrics_per_timestamp_.emplace_back(start, std::move(metrics));
+    }
+
+    const auto& end{std::chrono::system_clock::now()};
+    const auto& duration{end - start};
+    const auto& remainder{std::chrono::milliseconds(metrics_interval_ms_) -
+                          duration};
+
+    CheckForMetricIntervalTooShort(remainder, duration);
+
+    {
+      std::unique_lock<std::mutex> query_loop_lock{query_loop_mutex_};
+      query_loop_cv_.wait_for(query_loop_lock, remainder);
+    }
```

Comment:
Instead of using condition variable and mutex I think you can just use `sleep_for`:

https://en.cppreference.com/w/cpp/thread/sleep_for

Reply:
The condition variable allows us to wait but also wake-up early if profiling has completed. Imagine a situation where the metric interval is 4500 ms and the time window is 5000 ms. If the run finishes stabilizing on the third pass, then 15000 ms will have passed, but the metrics querying would still have 18000 ms mark to finish its 4th metric collection. In order to avoid wasting 3000 ms waiting for the last metric collection (because it is not needed since it's outside of the stability passes), using a condition variable here that can be woken up at the 15000 ms mark and querying loop ended is a time save for the user and better UX.

Reply:
I see. Thanks for the detailed response! Resolving.